pkg/kubelet/container/helpers.go

```diff
@@ -117,6 +117,22 @@ func HashContainer(container *v1.Container) uint64 {
 	return uint64(hash.Sum32())
 }
 
+func HashContainerForContainerChange(container *v1.Container) uint64 {
+	hash := fnv.New32a()
+	containerJSON, _ := json.Marshal(pickFieldToHash(container))
+	hashutil.DeepHashObject(hash, containerJSON)
+	return uint64(hash.Sum32())
+}
+
+// pickFieldToHash pick fields that will affect the running status of the container for hash,
+// currently this field range only contains `image`.
+func pickFieldToHash(container *v1.Container) *v1.Container {
+	return &v1.Container{
+		Name:  container.Name,
+		Image: container.Image,
+	}
+}
+
 // HashContainerWithoutResources returns the hash of the container with Resources field zero'd out.
 func HashContainerWithoutResources(container *v1.Container) uint64 {
```

Comment:
I think HashContainerWithoutResources should probably go away and HashContainerForContainerChange be made feature-aware so it includes the resources if the feature is enabled and does not include them if the feature is disabled

Reply:
Similarly, we only need to keep one hash function, because the intention of `HashContainerWithoutResources` is to prevent container `resource` field changes from restarting the container, and the intention of `HashContainerForContainerChange` (now the name has been changed to `HashContainer`) is the same.

The question now is, do we need to remove [HashWithoutResources](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/kubelet/container/runtime.go#L298-L302) and its related code? Because its content is already the same as [Hash](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/kubelet/container/runtime.go#L297), or is it to be kept until GA+1 as per the comments?

Reply:
@kubernetes/sig-node-pr-reviews Would appreciate any advice.

Reply:
> The question now is, do we need to remove [HashWithoutResources](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/kubelet/container/runtime.go#L298-L302) and its related code? Because its content is already the same as [Hash](https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/kubelet/container/runtime.go#L297), or is it to be kept until GA+1 as per the comments?

I think we should remove it, the single hash is informed by whether the kubelet has the feature enabled

Reply:
removed